UnitControl component: Refactor utils to TypeScript

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

Following on from discussion in https://github.com/WordPress/gutenberg/pull/34768#pullrequestreview-756120352, this PR refactors the UnitControl component's `utils` file and set of functions to TypeScript and adds a couple of types. To avoid creating a large PR that might make reviewing difficult, I thought we could look at refactoring this component one file at a time, with this one just focusing on the utils for the moment.

It looks like I could safely remove the lodash dependency from the file, too. I've also updated the documentation for some of the attributes where there were some slight inaccuracies.

Something I noticed while working on the file is that often the list of `units` could have a value set to `false` (there were a few places where this is expected). I believe this is to support either `useCustomUnits` returning the false value to switch off custom units, or possibly a `theme.json` file explicitly setting units to `false`. I _think_ I've managed to get this fairly consistent, but I'm very open to feedback if folks think there's a better way to deal with it.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

### Ensure tests pass

```
npm run test-unit -- --testPathPattern packages/components/src/unit-control/test/
```

### Ensure the UnitControl component works correctly

Via Storybook, run `npm run storybook:dev` and then visit http://localhost:50240/?path=/story/components-unitcontrol--default

Or, in the editor, go to edit a block that uses padding and adjust the controls (the Cover block is a good one to try)

### Ensure that a theme's theme.json units settings overrides the available units

With TT1-Blocks applied, you should see that the list of available units does not include `%`, but you can also manually edit the theme.json file's `settings.spacing.units` attribute to a reduced list of unit strings (e.g. `[ 'px', 'em', 'rem' ]` to ensure it works as expected.

## Screenshots <!-- if applicable -->

<img src="https://user-images.githubusercontent.com/14988353/134858805-46fb27d6-956d-4c00-b235-e9c3a76d3a6d.png" width=250 />

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Code quality, non-breaking change.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
UnitControl: Refactor utils to TypeScript

andrewserong commented:
CC @ciampo, I've started looking into refactoring the UnitControl component to TypeScript — I thought I'd try doing it in smaller PRs one file at a time to hopefully make it easier to review, but also since I don't have much TypeScript experience. Hopefully this will give me the chance to learn from any feedback before venturing too much deeper! 😄

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -0,0 +1,38 @@
+export interface WPUnitControlUnit {
```
We usually prefer `type` over `interface`

(here's an [example why](https://fettblog.eu/tidy-typescript-prefer-type-aliases/))

andrewserong replied:
Thanks for link, much appreciated, that makes perfect sense now, so I've updated this to `type` instead 👍

ciampo commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -1,29 +1,17 @@
-/**
- * External dependencies
- */
-import { isEmpty } from 'lodash';
```
Love it when we can rely on native features

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -0,0 +1,38 @@
+	step?: number;
+}
+
+export interface UseCustomUnitsProps {
```
Maybe we can find a better name for this type? Or maybe we could move the types directly to the function's declaration, and add the param comments via JSDoc.

andrewserong replied:
Good idea. I've moved it over to the function declaration instead of being a separate type (since there's only one place we'll use those props). Happy to move it around again if it looks messy, though!

ciampo replied:
That looks great, thank you!

ciampo commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -302,17 +309,17 @@ export const useCustomUnits = ( { units, availableUnits, defaultValues } ) => {
-	legacyUnit,
-	units = ALL_CSS_UNITS
-) {
+	currentValue: number | string,
```
I would probably add a new type in `types.ts`:

```ts
type Value = number | string;
```

And then replace it everywhere we currently intend to represent such a value (including in the `WPUnitControlUnit` and `UseCustomUnitsProps` types).

andrewserong replied:
I like that idea, it neatens things a fair bit and ensures consistency (previously I see I'd inconsistently used `number | string` and `string | number`). I've updated, but please let me know if I've missed anything.

ciampo commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -302,17 +309,17 @@ export const useCustomUnits = ( { units, availableUnits, defaultValues } ) => {
-) {
+	currentValue: number | string,
+	legacyUnit: string | undefined,
+	units: Array< WPUnitControlUnit > | false = ALL_CSS_UNITS
+): Array< WPUnitControlUnit > | false {
```
Using `false` as a valid value for `units` and also as a potential return value of the function feels a bit weird (I know the implementation was there before this PR, and in fact, it's thanks to using TypeScript that we can spot situations like this one).

I can see that the same pattern happens in the `hasUnits` function — this is probably something that we should tackle separately in a follow-up PR. WDYT?

andrewserong replied:
I thought the same, it does feel a bit weird in these functions. I _think_ we'll still need to support `theme.json` providing `false` as a valid value to switch off units support, but there's likely a neater way of handling it than having these functions use it as an expected type.

> this is probably something that we should tackle separately in a follow-up PR.

Agreed, it'll probably be easier to do as a follow-up as the implementation might need to change slightly.

ciampo commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -210,18 +205,23 @@ export function parseUnit( initialValue, units = ALL_CSS_UNITS ) {
+ * @param  units         Units to derive from.
+ * @param  fallbackValue The fallback value.
+ * @param  fallbackUnit  The fallback value.
+ * @return The extracted number and unit.
```
```suggestion
 * @return The extracted value and unit.
```

ciampo commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -210,18 +205,23 @@ export function parseUnit( initialValue, units = ALL_CSS_UNITS ) {
+	let baseUnit: string;
 
-	if ( isNaN( parsedValue ) || parsedValue === '' ) {
+	if ( ! Number.isFinite( parsedValue ) || parsedValue === '' ) {
```
We should consider this change carefully, since it introduces a runtime change — `isNaN` doesn't always return the same results as `! Number.isFinite`

andrewserong replied:
Thanks for pointing this out, I meant to leave a specific comment for it. With the switch to TypeScript, using `isNaN` here generated an error:

![image](https://user-images.githubusercontent.com/14988353/135031441-6c4c17a4-b268-4972-95cc-e91bfb5e0eb6.png)

This seems to be that in TS, `isNaN` expects to be called with a variable of type `number`, but our `parsedValue` is either a number or a string. At this point in `getValidParsedUnit` the `parsedValue` variable must be either a number or an empty string (the results of calling `parseUnit`). So, I think `Number.isFinite` is a more appropriate check than `isNaN` anyway, at least from my mental model of what this function should be doing:

* Check that the number is not `NaN`
* Check that the number is a valid value that we can use (e.g. not `Infinity` or `-Infinity`) and not an arbitrary string

If it doesn't meet the above criteria then we fall back to the `fallbackValue`. One difference in using `Number.isFinite` is that the value won't be coerced to a number before checking to see if it's finite. I _think_ this is okay, because the value has already been coerced in `parseUnit` so at this point it should be a real (finite) value in order for us to consider it valid.

Please let me know if you think that's not right, though! Happy to change the logic here, I think it's the main place where I had to introduce a runtime change, but 🤞 it isn't an unexpected one in usage.

It could be that we might want to tighten up the Value type a bit further, too, given that in some places it'll only be `number` or the string literal `''` — I was a bit cautious about doing that, as we'd then have two different Value types floating around in the file, which could get confusing? But it also might make it a bit clearer 🤔

ciampo replied:
> Please let me know if you think that's not right, though! Happy to change the logic here, I think it's the main place where I had to introduce a runtime change, but 🤞 it isn't an unexpected one in usage.

Thank you for the explanation! With this context (and the extra reassurance given by the passing unit tests), I think we can leave your new implementation untouched.

I would maybe only add a **comment** with a short version of the explanation you just gave in your previous message.

> It could be that we might want to tighten up the Value type a bit further, too, given that in some places it'll only be number or the string literal '' — I was a bit cautious about doing that, as we'd then have two different Value types floating around in the file, which could get confusing? But it also might make it a bit clearer 🤔

I had the same thought and came to the same conclusion

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -0,0 +1,38 @@
+	/**
+	 * The value for the unit, used in a CSS value (e.g `px`).
+	 */
+	value: string;
```
Should we look into typing the allowed units more strictly? I'm not fully aware if users of `UnitControl` are allowed to specify just about any unit they want.

andrewserong replied:
That's a good question. I think when it comes to the value in `WPUnitControlUnit` a case can be made for allowing arbitrary values so that the UnitControl component can be used for units that we haven't factored in (or thought of). An example came to mind — if someone were to develop a Map block for real estate listings or something of that nature, they might want to include a UnitControl for adjusting the search radius (or zoom level) in the map. In that case, perhaps they'd use units like `mi` and `km` for distances.

I see there's already a test for custom unit values, but I've added in an additional storybook story that includes custom units for `km`, `mi`, `m`, and `yd` as an example, so that we've got it in the component.

ciampo replied:
Thank you for the explanation, and for adding a Storybook example. Let's leave it as a more relaxed `string` type

ciampo reviewed (commented):
Thank you for your work on this, @andrewserong, it is much needed!

I've left a few comments inline, but overall it's looking good!

andrewserong committed:
Update return value doc comment for getValidParsedUnit

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

andrewserong committed:
Adjust types, add story for custom units

andrewserong committed:
Use Value in more places for consistency

andrewserong committed:
Update default attribute on WPUnitControlUnit to use Value type

andrewserong commented:
Thanks for the detailed feedback @ciampo, much appreciated! I think I've either answered or implemented all the feedback and questions, but do let me know if there are any other changes you'd like to see 🙂

ciampo commented on packages/components/src/unit-control/utils.ts:
```diff
@@ -171,32 +163,35 @@ export function getParsedValue( value, unit, units ) {
  */
-export function hasUnits( units ) {
-	return ! isEmpty( units ) && units !== false;
+export function hasUnits( units: Array< WPUnitControlUnit > | false ): boolean {
```
We could also extract `Array< WPUnitControlUnit > | false` type to `types.ts`, to make this pattern more clear. Maybe something like

```ts
type WPUnitControlUnitList = Array< WPUnitControlUnit > | false;
```

This would also help us when potentially working on https://github.com/WordPress/gutenberg/pull/35138#discussion_r717246773

ciampo reviewed (commented):
Most feedback has been addressed, I just left a couple of minor comments. This is close to getting merged!

ciampo reviewed (approved):
Actually, to speed this up, I can already approve this PR, since the two actionable pieces of feedback that I left are quite trivial and wouldn't change the runtime code. Feel free to address them and merge! 🚀 

---

I can see two follow-ups from here:

- Fully type UnitControl. If you're in search of good examples on how to do this, #34028 may be of help
- Investigate how to potentially improve the types around units as discussed in https://github.com/WordPress/gutenberg/pull/35138#discussion_r717246773

I'm happy to help with both items, in case you have the time to work on them!

andrewserong committed:
Add WPUnitControlUnitList type

andrewserong committed:
Add contextual comment to the Number.isFinite check

andrewserong commented:
Thanks @ciampo! I've added in the contextual comment for the `isFinite` check, and the additional type.

I did add in another small runtime change to `filterUnitsWithSettings`. When adding in the new `WPUnitControlUnitList` type, I noticed that this function expected to be called with an array. In all current usage of it, it _is_ called with an array, but given that the other similar functions are written to also factor in that they could be called with `false`, I've added in another `Array.isArray` check, with a fallback to an empty array, to harden the function a little. I've also included an additional unit test.

I've tested the change in a couple of different themes (with and without `theme.json` files with custom units settings) and in the iOS simulator to make sure that the native Button block border radius control still uses its filtered units correctly. So, once the performance tests pass, I'll merge this one in.

Happy to continue work in follow-ups (I'll look at fully typing the rest of the UnitControl component later on this week), but let me know if there are any issues and I can do other follow-ups too.

ciampo commented:
Thank you!

In the context of refactoring to the `WPUnitControlUnitList` type, I noticed that the `units` argument in the `getParsedValue`, `filterUnitsWithSettings` and `useCustomUnits` functions didn't previously accept `false` — it's probably a good idea to keep it the way it used to be, so that we don't widen the types more than they need to be?

We could make these changes in the follow-up issue/PR that investigates how to improve the `WPUnitControlUnitList` type

andrewserong commented:
Thanks again @ciampo!

> We could make these changes in the follow-up issue/PR that investigates how to improve the `WPUnitControlUnitList` type

Yes, let's look at it in that context in the follow-up. Hopefully it should be a little clearer / cleaner when we're just looking at that one thing in isolation.

I thought I'd start on typing the JSX -> TSX first later on this week, and then come back to the `WPUnitControlUnitList` type, but let me know if you'd prefer we do it the other way around 🙂

ciampo commented:
> I thought I'd start on typing the JSX -> TSX first later on this week, and then come back to the `WPUnitControlUnitList` type, but let me know if you'd prefer we do it the other way around 🙂

Sounds good to me! Feel free to open a draft issue and ping me there if you want early feedback too :)
